File: src/libs/m3/session/M3FS.cc
/*
 * Copyright (C) 2016-2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <m3/com/GateStream.h>
#include <m3/session/M3FS.h>
#include <m3/vfs/GenericFile.h>

namespace m3 {

File *M3FS::open(const char *path, int perms) {
    capsel_t ep;
    if((perms & FILE_NOSESS) && (ep = alloc_ep()) != ObjCap::INVALID) {
        GateIStream reply = send_receive_vmsg(_gate, OPEN_PRIV, path, perms, ep - _eps);
        reply >> Errors::last;
        if(Errors::last != Errors::NONE)
            return nullptr;
        size_t id;
        reply >> id;
        return new GenericFile(perms, sel(), id, VPE::self().sel_to_ep(ep), this);
    }
    else {
        perms &= ~FILE_NOSESS;
        KIF::ExchangeArgs args;
        args.count = 1;
        args.svals[0] = static_cast<xfer_t>(perms);
        strncpy(args.str, path, sizeof(args.str));
        KIF::CapRngDesc crd = obtain(2, &args);
        if(Errors::last != Errors::NONE)
            return nullptr;
        return new GenericFile(perms, crd.start());
    }
}

Errors::Code M3FS::stat(const char *path, FileInfo &info) {
    GateIStream reply = send_receive_vmsg(_gate, STAT, path);
    reply >> Errors::last;
    if(Errors::last != Errors::NONE)
        return Errors::last;
    reply >> info;
    return Errors::NONE;
}

Errors::Code M3FS::mkdir(const char *path, mode_t mode) {
    GateIStream reply = send_receive_vmsg(_gate, MKDIR, path, mode);
    reply >> Errors::last;
    return Errors::last;
}

Errors::Code M3FS::rmdir(const char *path) {
    GateIStream reply = send_receive_vmsg(_gate, RMDIR, path);
    reply >> Errors::last;
    return Errors::last;
}

Errors::Code M3FS::link(const char *oldpath, const char *newpath) {
    GateIStream reply = send_receive_vmsg(_gate, LINK, oldpath, newpath);
    reply >> Errors::last;
    return Errors::last;
}

Errors::Code M3FS::unlink(const char *path) {
    GateIStream reply = send_receive_vmsg(_gate, UNLINK, path);
    reply >> Errors::last;
    return Errors::last;
}

Errors::Code M3FS::delegate(VPE &vpe) {
    if(vpe.delegate_obj(sel()) != Errors::NONE)
        return Errors::last;
    // TODO what if it fails?
    return obtain_for(vpe, KIF::CapRngDesc(KIF::CapRngDesc::OBJ, sel() + 1, 1));
}

void M3FS::serialize(Marshaller &m) {
    m << sel();
}

FileSystem *M3FS::unserialize(Unmarshaller &um) {
    capsel_t sel;
    um >> sel;
    return new M3FS(sel);
}

}

File: src/libs/rustm3/src/session/m3fs.rs
/*
 * Copyright (C) 2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use cap::Selector;
use cell::RefCell;
use col::Vec;
use com::*;
use core::any::Any;
use core::{fmt, intrinsics};
use errors::Error;
use kif;
use rc::{Rc, Weak};
use serialize::Sink;
use session::ClientSession;
use util;
use vfs::{FileHandle, FileInfo, FileMode, FileSystem, FSHandle, GenericFile, OpenFlags};
use vpe::VPE;

pub type ExtId = u16;

pub struct M3FS {
    self_weak: Weak<RefCell<M3FS>>,
    sess: ClientSession,
    sgate: SendGate,
}

int_enum! {
    struct Operation : u32 {
        const STAT      = 0x5;
        const MKDIR     = 0x6;
        const RMDIR     = 0x7;
        const LINK      = 0x8;
        const UNLINK    = 0x9;
    }
}

bitflags! {
    pub struct LocFlags : u32 {
        const BYTE_OFF  = 0b01;
        const EXTEND    = 0b10;
    }
}

impl M3FS {
    fn create(sess: ClientSession, sgate: SendGate) -> FSHandle {
        let inst = Rc::new(RefCell::new(M3FS {
            self_weak: Weak::new(),
            sess: sess,
            sgate: sgate,
        }));
        inst.borrow_mut().self_weak = Rc::downgrade(&inst);
        inst
    }

    pub fn new(name: &str) -> Result<FSHandle, Error> {
        let sels = VPE::cur().alloc_sels(2);
        let sess = ClientSession::new_with_sel(name, 0, sels + 1)?;

        let crd = kif::CapRngDesc::new(kif::CapType::OBJECT, sels + 0, 1);
        let mut args = kif::syscalls::ExchangeArgs::default();
        sess.obtain_for(VPE::cur().sel(), crd, &mut args)?;
        let sgate = SendGate::new_bind(sels + 0);
        Ok(Self::create(sess, sgate))
    }

    pub fn new_bind(sels: Selector) -> FSHandle {
        Self::create(ClientSession::new_bind(sels + 0), SendGate::new_bind(sels + 1))
    }

    pub fn sess(&self) -> &ClientSession {
        &self.sess
    }
}

impl FileSystem for M3FS {
    fn as_any(&self) -> &Any {
        self
    }

    fn open(&self, path: &str, flags: OpenFlags) -> Result<FileHandle, Error> {
        let mut args = kif::syscalls::ExchangeArgs {
            count: 1,
            vals: kif::syscalls::ExchangeUnion {
                s: kif::syscalls::ExchangeUnionStr {
                    i: [flags.bits() as u64, 0],
                    s: unsafe { intrinsics::uninit() },
                },
            },
        };

        // copy path
        unsafe {
            for (a, c) in args.vals.s.s.iter_mut().zip(path.bytes()) {
                *a = c as u8;
            }
            args.vals.s.s[path.len()] = '\0' as u8;
        }

        let crd = self.sess.obtain(2, &mut args)?;
        Ok(Rc::new(RefCell::new(GenericFile::new(crd.start()))))
    }

    fn stat(&self, path: &str) -> Result<FileInfo, Error> {
        let mut reply = send_recv_res!(
            &self.sgate, RecvGate::def(),
            Operation::STAT, path
        )?;
        Ok(reply.pop())
    }

    fn mkdir(&self, path: &str, mode: FileMode) -> Result<(), Error> {
        send_recv_res!(
            &self.sgate, RecvGate::def(),
            Operation::MKDIR, path, mode
        ).map(|_| ())
    }
    fn rmdir(&self, path: &str) -> Result<(), Error> {
        send_recv_res!(
            &self.sgate, RecvGate::def(),
            Operation::RMDIR, path
        ).map(|_| ())
    }

    fn link(&self, old_path: &str, new_path: &str) -> Result<(), Error> {
        send_recv_res!(
            &self.sgate, RecvGate::def(),
            Operation::LINK, old_path, new_path
        ).map(|_| ())
    }
    fn unlink(&self, path: &str) -> Result<(), Error> {
        send_recv_res!(
            &self.sgate, RecvGate::def(),
            Operation::UNLINK, path
        ).map(|_| ())
    }

    fn fs_type(&self) -> u8 {
        b'M'
    }
    fn exchange_caps(&self, vpe: Selector,
                            dels: &mut Vec<Selector>,
                            max_sel: &mut Selector) -> Result<(), Error> {
        dels.push(self.sess.sel());

        let crd = kif::CapRngDesc::new(kif::CapType::OBJECT, self.sess.sel() + 1, 1);
        let mut args = kif::syscalls::ExchangeArgs::default();
        self.sess.obtain_for(vpe, crd, &mut args)?;
        *max_sel = util::max(*max_sel, self.sess.sel() + 2);
        Ok(())
    }
    fn serialize(&self, s: &mut VecSink) {
        s.push(&self.sess.sel());
        s.push(&self.sgate.sel());
    }
}

impl M3FS {
    pub fn unserialize(s: &mut SliceSource) -> FSHandle {
        let sels: Selector = s.pop();
        M3FS::new_bind(sels)
    }
}

impl fmt::Debug for M3FS {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "M3FS[sess={:?}, sgate={:?}]", self.sess, self.sgate)
    }
}
